tools/launch_instance.rb

```diff
@@ -885,6 +882,17 @@ def terminate(vminfo)
             config: vars_file.path,
             launched_instances_name_prefix: spec[:name_prefix]
           )
+        when /^dns_component/
```

Comment:
Per my testing, https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/ocp-common/job/Flexy-install/191932/artifact/vminfo.yml/*view*/, the key is "**:**dns_component", I am a bit concerned if it can be matched.

Reply:
@jianlinliu even though the key is a symbol is the YAML when parsing it, you can try it as a String when used with the `case` (you can try the code below to verify).  As you can see from the run log output, this parsing does work.  IMO, it's better than [your approach of deleting the element ](https://github.com/openshift/verification-tests/pull/3284/files#diff-d212a2448388579d50551d687d8495db9858e3e253a7a1c26814085acdd51df8R860-R895)
```
require 'yaml'
template_yaml = """
---
template_random:
  :config: private-templates/functionality-testing/aos-4_12/ipi-on-aws/versioned-installer
  :template: template-destroy-ipi-versioned
  :name_prefix: jialiu1
:dns_component: 0403os3-yf1
"""

vminfo = YAML.load(template_yaml)

vminfo.each do |target, spec|
  case target
  when /^template_/
    print("TARGET: #{target}\n")
  when /^dns_component/
    print("##### GOT DNS_COMPONENT\n")
  else
    print("XXXX fall-through...\n")

  end
end
```

Reply:
> even though the key is a symbol is the YAML when parsing it, you can try it as a String when used with the case

Good to know.

Some other reason to use the approach of deleting the element, if place `dns_component` as a dedicated case here, then whatever 4.x or 3.x install, `delete_resource_records_re` is always executed to call aws api (`list_resource_record_sets`), even it return empty results.

Reply:
@jianlinliu your approach is making an exception by removing the element `dns_component` to avoid it being processed by the `vminfo` loop but later on processes it, which might make it harder to understand.  You are correct in that it will be executed regardless there's a `dns_component` or not, I just feel that having a patterned workflow is preferable over making exceptions.  @gpei @liangxia @jhou1, mind giving your input on this decision?  I'm ok with either way, but would like to get some other perspectives.

Reply:
I don't fully understand both, based on my limited knowledge. 

I'd suggest not altering the value of the input parameter `vminfo` for maintainability.

I'd also suggest creating a method such as `delete_aws_dns_records` and then in `terminate` call that method conditionally to keep the method of `terminal` shorter.

The[ yaml content generated](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/ocp-common/job/Flexy-install/191932/artifact/vminfo.yml/*view*/) has a colon prefixing the key, prefer to have the colon removed for readability to people building such jobs.

Reply:
@jhou1 the deletion portion is only 3 lines (2 if we fold the definition of the `Regexp` into the function call itself).  The YAML element parsing line `when /^dns_component/` works for both `String` or `Symbol`, so we are covered on that front.  Regarding invoking AWS API, we have to use it to remove the resources...this is better than calling AWS CLI command directly, which ultimately is a wrapper around AWS API anyway.